### ISSUE TITLE
Can't tab out of field when calendar is open and breaks accessibility

### ISSUE THREAD

**Issue by jwallet** [Hacker0x01/react-datepicker#2139]
**Describe the bug**
2.9.6 was not crashing using Select Year/Month dropdown
2.15.0 fixed it but it is not possible to tab out of the field and get the value selected when moving to the next field 2.9.6 does work.

**To Reproduce**

1. Focus on the field
2. use keyboards arrows to move to a date without selecting it
3. hit tab or shift+tab to move to another field
4. it does not move to the other field, it focus inside the calendar and it also breaks the accessibilty hotkeys

- Firefox
- Chrome

**Comment by ajcrews** [comment]
Looks related to https://github.com/Hacker0x01/react-datepicker/issues/2140. Even a user that has a mouse can't access the other date field without double-clicking. This is a breaking change and we are locking down the version of this package in our repos until it is fixed.

**Comment by akash87** [comment]
Any workarounds for this ?

**Comment by bgl-tand** [comment]
This can solve half of the issue to set the value on tab. Otherwise just use the older version.
`onKeyDown = (e: any) => {
        if (e.key == 'Tab') {
            setTimeout(() =>
                this.ref!.setOpen(false), 100
            )
        }
    }`

Only work around for moving to the next field i've found so far is to specify where you want to focus next, but quite tedious to have to specify this everywhere.

**Comment by mtran-method** [comment]
Hi, I'm still experiencing this issue when a timepicker is present with version 3.4.1 

Unable to tab out and breaking accessbility.

**Comment by reinvanimschoot** [comment]
I have the same issue, any updates on this?

**Comment by bryn-heath** [comment]
Still got this issue   - Win7 + IE 11

**Comment by GalGrigoryeva** [comment]
Will this issue be fixed?

**Comment by jp-albrecht** [comment]
I'm experiencing this issue as well. Any update?

**Comment by avi-dwave** [comment]
This is an issue for me as well

**Comment by kiprobinson** [comment]
> This can solve half of the issue to set the value on tab. Otherwise just use the older version.
> `onKeyDown = (e: any) => { if (e.key == 'Tab') { setTimeout(() => this.ref!.setOpen(false), 100 ) } }`
> 
> Only work around for moving to the next field i've found so far is to specify where you want to focus next, but quite tedious to have to specify this everywhere.
> 
> I ended up going with version 2.9.0. This solves any tabbing issues in my case and requires no workaround code - keeping things simple.

2.10.1 worked for me, and it is definitely broken by 2.16.0. I'm not going to go version by version to try to figure out which one broke it. Still broken in latest (4.1.1 as I write this).

Also if you tab into the field, type in a date manually, then tab out, the tab gets stuck looping through prev month, next month, and date. It won't tab to the next field ever.

**Comment by tammypmc** [comment]
Hello! I need to know if this issue still present on the current version?

**Comment by kiprobinson** [comment]
@tammypmc 

> Hello! I need to know if this issue still present on the current version?

I'm no longer working on any react project. But I can reproduce on the demo page: https://reactdatepicker.com/

1. Click near the input field on the first example (not the one at the top, it does other weird stuff. The first one under "Examples".)
2. Tab => Focus goes into the input field
3. Start hitting tab again => It loops between input field, prev month, next month, selected date, and back to input field.

I'm on Chrome on Windows. I've attached a video - all I'm doing in the video is hitting tab over and over

https://user-images.githubusercontent.com/2453426/135771190-288484e2-ed02-4396-8072-8567545da3d0.mp4

**Comment by tperamaki** [comment]
Steps to produce: Click tab until you are stuck in a loop
Requirements for it to happen: There needs to be the arrows for changing months visible

So if you use `includeDates` option to limit the dates to only one month, there won't be any arrows visible, and the user can properly tab out of the datepicker (although it's left open).

**Comment by tperamaki** [comment]
There is a tab loop feature in this library that is by default enabled, which breaks the accessibility and works against the documentation (which states that tab exits). See https://github.com/Hacker0x01/react-datepicker/blob/18302ebb2cb6b98ae9064e195d73bc25e96eb6a9/src/tab_loop.jsx

Thankfully the workaround is rather easy, just add the property `enableTabLoop={false}` and you can tab out of it again, although it will stay open, but still better.

```
<DatePicker enableTabLoop={false} selected={startDate} onChange={(date) => setStartDate(date)} />
```

**Comment by shanebeehler** [comment]
My solution is to disable the calendar opening on focus which will allow tabbing past this input. And then allowing the user to hit Enter key to open the calendar when on the input. From there, the user can select a date via Enter key or close the calendar via Esc key to the tab on.

```
  const handleOnKeyDownDatePicker = event => {
    if (datePickerRef?.current && !datePickerRef.current.state.open) {
      if (event?.keyCode === 13 || event?.which === 13) {
        event.preventDefault();
        datePickerRef.current?.setOpen(true);
      }
    }
  };
  
  <ReactDatePicker
    ...
    preventOpenOnFocus={true}
    ref={datePickerRef}
    onKeyDown={handleOnKeyDownDatePicker}
  />
```

**Comment by ryanmunger** [comment]
Anymore information on this one? It would be great to have at least an option to cycle to the next input after the calendar closes. Seeing the issue when a user tabs in, focuses, selects a date. Hitting tab again just selects the same input over and over again.

**Comment by ch1qui** [comment]
This is not an issue, Check out the modal pattern as explained in this WAI-ARIA article:
https://www.w3.org/WAI/ARIA/apg/patterns/dialogmodal/
_If focus is on the last tabbable element inside the dialog, moves focus to the first tabbable element inside the dialog._
Since the tooltip is interactive, having a focus-trap is a good thing, its not breaking accesibility.

**Comment by kiprobinson** [comment]
> This is not an issue, Check out the modal pattern as explained in this WAI-ARIA article:

Have you tried it though? The user is literally incapable of going anywhere else on the page without using the mouse. Hitting enter or space bar doesn't dismiss the dialog, neither does escape key.

**Comment by ryanmunger** [comment]
@ch1qui this issue happens after you have closed the calendar and selected a date. So the flow goes:

- tab into a field
- Calendar pops
- Choose date
- Calendar closes
- Tab into next field

where I think it’s broken is the fact that it pops the SAME calendar on the SAME field previously selected. Therefore a keyboard user can never tab to the next field. Hope this makes more sense!

**Comment by SeanDonnellanJH** [comment]
When I try to skip over the calendar entirely with tab there is no way to do so, when I tab into the input button and then press tab again would expect to focus on the next input but instead the calendar dropdown opens automatically (I would have expected it to open like a dropdown via the space bar or literally anything other than tab). Then when I am in the component, I tab through it all and it loops, as expected. But when I press escape my focus is brought back to the input field for the date picker, which as I explained I cannot skip.

How is a keyboard user expected to access content past your component?

**Comment by martijnrusschen** [comment]
Yes, I agree @reinrl. We shouldn't treat this as a modal but more as a native part of the page you're on. Users should be able to tab out of it like other fields. However, looking at the examples here: https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/datepicker-dialog/ it seems the recommended path is to treat a datepicker as a modal. I'm not 100% sure what's best since I'm not an accessibility expert.

**Comment by reinrl** [comment]
@martijnrusschen Thanks for the quick reply - I stand corrected (sorry @ch1qui - that guidance from the WAI really surprises me, based on what I thought that I knew about other various input elements!). I don't profess to be an A11Y expert either (only having benefited from working with a few different vendors at various jobs), so our next steps are going to be to go back to our current compliance vendor with that link in hand, and see what they have to say about that information versus their own guidance that the calendar view should not trap the user's focus. We usually hear back pretty quickly, so I will circle back here once I hear more.

**Comment by martijnrusschen** [comment]
Sounds good!

**Comment by reinrl** [comment]
> @martijnrusschen Thanks for the quick reply - I stand corrected (sorry @ch1qui - that guidance from the WAI really surprises me, based on what I thought that I knew about other various input elements!). I don't profess to be an A11Y expert either (only having benefited from working with a few different vendors at various jobs), so our next steps are going to be to go back to our current compliance vendor with that link in hand, and see what they have to say about that information versus their own guidance that the calendar view should not trap the user's focus. We usually hear back pretty quickly, so I will circle back here once I hear more.

My apologies, as it would appear that I never did circle back on this in a timely fashion. After some further conversation with our A11Y vendor, we settled on a bit of a hybrid approach:
- when open, the date picker does act like a modal (trapping focus within the various actions presented)
- we no longer open the date picker "modal" when the field is focused (instead adding a calendar icon button to trigger the opening), so you can continue to tab right through the fields in the form without getting trapped on a date field

![image](https://github.com/user-attachments/assets/3d72aee8-8ae6-4482-8435-bac71aad155a)
